### ISSUE TITLE
kube-controller-manager crashed with fatal error: concurrent map writes

### ISSUE THREAD

**Issue by davidwtf** [kubernetes/kubernetes#121053]
### What happened?

The kube-controller-manage may crash when processing StatefulSet with spec.podManagementPolicy is Parallel.
With logs:
```
......
I1008 08:06:57.041752       1 event.go:307] "Event occurred" object="default/web" fieldPath="" kind="StatefulSet" apiVersion="apps/v1" type="Normal" reason="SuccessfulCreate" message="create Pod web-90 in StatefulSet web successful"
I1008 08:06:57.076495       1 event.go:307] "Event occurred" object="default/web" fieldPath="" kind="StatefulSet" apiVersion="apps/v1" type="Normal" reason="SuccessfulCreate" message="create Pod web-66 in StatefulSet web successful"
fatal error: concurrent map writes

goroutine 5844 [running]:
k8s.io/kubernetes/pkg/controller/statefulset.getPersistentVolumeClaims(0xc003ee8500, 0x0?)
        pkg/controller/statefulset/stateful_set_utils.go:348 +0x2fd
k8s.io/kubernetes/pkg/controller/statefulset.(*StatefulPodControl).createPersistentVolumeClaims(0xc000b24560, 0x6ecea2?, 0xc000600000?)
        pkg/controller/statefulset/stateful_pod_control.go:341 +0x6a
k8s.io/kubernetes/pkg/controller/statefulset.(*StatefulPodControl).createMissingPersistentVolumeClaims(0x4a1f120?, {0x4a168a0, 0xc000422f50}, 0x448b653?, 0x36?)
        pkg/controller/statefulset/stateful_pod_control.go:321 +0x3f
k8s.io/kubernetes/pkg/controller/statefulset.(*defaultStatefulSetControl).processReplica(0xc000f44ec0, {0x4a168a0, 0xc000422f50}, 0xc003ee8500, 0xc003e383a0, 0xc003e383a0, 0xc00583df60?, 0x1c7c4c6?, 0x0, {0xc005114a80, ...}, ...)
        pkg/controller/statefulset/stateful_set_control.go:420 +0x68c
k8s.io/kubernetes/pkg/controller/statefulset.(*defaultStatefulSetControl).updateStatefulSet.func1(0x4a168a0?)
        pkg/controller/statefulset/stateful_set_control.go:626 +0x65
k8s.io/kubernetes/pkg/controller/statefulset.slowStartBatch.func1(0x0?)
        pkg/controller/statefulset/stateful_set_control.go:292 +0x68
created by k8s.io/kubernetes/pkg/controller/statefulset.slowStartBatch
        pkg/controller/statefulset/stateful_set_control.go:289 +0x16a

goroutine 1 [chan receive (nil chan)]:
k8s.io/kubernetes/cmd/kube-controller-manager/app.Run({0x4a168d8, 0xc00013a000}, 0xc000624760)
        cmd/kube-controller-manager/app/controllermanager.go:330 +0x10e5
k8s.io/kubernetes/cmd/kube-controller-manager/app.NewControllerManagerCommand.func2(0xc000c22600?, {0xc00089f800?, 0x0?, 0x17?})
        cmd/kube-controller-manager/app/controllermanager.go:144 +0x21c
github.com/spf13/cobra.(*Command).execute(0xc000c22600, {0xc00012e010, 0x17, 0x17})
        vendor/github.com/spf13/cobra/command.go:916 +0x862
github.com/spf13/cobra.(*Command).ExecuteC(0xc000c22600)
        vendor/github.com/spf13/cobra/command.go:1040 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
        vendor/github.com/spf13/cobra/command.go:968
k8s.io/component-base/cli.run(0xc000c22600)
        vendor/k8s.io/component-base/cli/run.go:146 +0x317
k8s.io/component-base/cli.Run(0xc0000061a0?)
        vendor/k8s.io/component-base/cli/run.go:46 +0x1d
main.main()
        cmd/kube-controller-manager/controller-manager.go:36 +0x1e
 ......
```

### What did you expect to happen?

we expect the kube-controller-manage will work well.

### How can we reproduce it (as minimally and precisely as possible)?

The following example can reproduce this problem:
```
---
apiVersion: v1
kind: Service
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  ports:
  - port: 80
    name: web
  clusterIP: None
  selector:
    app: nginx
---
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: web
spec:
  serviceName: "nginx"
  replicas: 100
  podManagementPolicy: Parallel
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:alpine
        ports:
        - containerPort: 80
          name: web
        volumeMounts:
        - name: www
          mountPath: /usr/share/nginx/html
  volumeClaimTemplates:
  - metadata:
      name: www
      labels:
        a: b
    spec:
      accessModes: [ "ReadWriteOnce" ]
      resources:
        requests:
          storage: 1Gi
```

### Anything else we need to know?

After digging into the source code, we found that when using the `Parallel` pod management policy, the statefulset controller will use `slowStartBatch` to start parallel processing.
But the `getPersistentVolumeClaims` which is called during parallel processing may directly update the claims' labels without a lock. This finally causes a concurrent map-accessing problem.

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.5", GitCommit:"93e0d7146fb9c3e9f68aa41b2b4265b2fcdb0a4c", GitTreeState:"clean", BuildDate:"2023-08-24T00:48:26Z", GoVersion:"go1.20.7", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v5.0.1
Server Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.5", GitCommit:"93e0d7146fb9c3e9f68aa41b2b4265b2fcdb0a4c", GitTreeState:"clean", BuildDate:"2023-08-24T00:42:11Z", GoVersion:"go1.20.7", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
$ uname -a
# paste output here
Linux ait-dev-m1 5.4.240-1.el7.elrepo.x86_64 #1 SMP Tue Apr 4 13:24:42 EDT 2023 x86_64 x86_64 x86_64 GNU/Linux

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by liggitt** [comment]
Thanks for the report. Do we know when this issue was introduced?

cc @soltysh @aleksandra-malinowska

**Comment by nayihz** [comment]
Maybe it was imported by this pr https://github.com/kubernetes/kubernetes/pull/117865

**Comment by nayihz** [comment]
We can reproduce this issue in unit testing.
```go
diff --git a/pkg/controller/statefulset/stateful_set_control_test.go b/pkg/controller/statefulset/stateful_set_control_test.go
index f018646d3af..ea70ab34a97 100644
--- a/pkg/controller/statefulset/stateful_set_control_test.go
+++ b/pkg/controller/statefulset/stateful_set_control_test.go
@@ -2989,6 +2989,18 @@ func parallelScaleUpStatefulSetControl(set *apps.StatefulSet,
        if err != nil {
                return err
        }
+       rand.Seed(time.Now().UnixNano())
+
+       for i := 0; i < 100; i++ {
+               go func() {
+                       time.Sleep(time.Duration(rand.Int63n(201)) * time.Millisecond)
+                       if set.Labels != nil {
+                               set.Labels["aaaaa"] = "bbbbb"
+                       } else {
+                               fmt.Println("set.Labels is nil")
+                       }
+               }()
+       }

        // Give up after 2 loops.
        // 2 * 500 pods per loop = 1000 max pods <- this should be enough for all test cases.
diff --git a/pkg/controller/statefulset/stateful_set_utils_test.go b/pkg/controller/statefulset/stateful_set_utils_test.go
index d5d87724dad..630e8b6194c 100644
--- a/pkg/controller/statefulset/stateful_set_utils_test.go
+++ b/pkg/controller/statefulset/stateful_set_utils_test.go
@@ -840,6 +840,7 @@ func newStatefulSetWithVolumes(replicas int32, name string, petMounts []v1.Volum
                        Name:      name,
                        Namespace: v1.NamespaceDefault,
                        UID:       types.UID("test"),
+                       Labels:    make(map[string]string),
                },
                Spec: apps.StatefulSetSpec{
                        Selector: &metav1.LabelSelector{
```
```
root@dev:~/work/kubernetes/pkg/controller/statefulset# go test -v -run TestParallelScale
=== RUN   TestParallelScale
=== RUN   TestParallelScale/scale_up_from_3_to_30
=== RUN   TestParallelScale/scale_down_from_10_to_1
=== RUN   TestParallelScale/scale_down_to_0
fatal error: concurrent map writes

goroutine 57 [running]:
k8s.io/kubernetes/pkg/controller/statefulset.parallelScaleUpStatefulSetControl.func1()
	/root/work/kubernetes/pkg/controller/statefulset/stateful_set_control_test.go:2998 +0x5d
created by k8s.io/kubernetes/pkg/controller/statefulset.parallelScaleUpStatefulSetControl in goroutine 148
	/root/work/kubernetes/pkg/controller/statefulset/stateful_set_control_test.go:2995 +0x145

goroutine 1 [chan receive]:
testing.(*T).Run(0xc000580b60, {0x211051f?, 0x52fbbc?}, 0x2217768)
	/usr/local/go/src/testing/testing.go:1649 +0x3c8
testing.runTests.func1(0x362dd00?)
	/usr/local/go/src/testing/testing.go:2054 +0x3e
testing.tRunner(0xc000580b60, 0xc000641c48)
	/usr/local/go/src/testing/testing.go:1595 +0xff
testing.runTests(0xc000020f00?, {0x3616620, 0x57, 0x57}, {0x418a5f?, 0xc000641d08?, 0x362c520?})
	/usr/local/go/src/testing/testing.go:2052 +0x445
testing.(*M).Run(0xc000020f00)
	/usr/local/go/src/testing/testing.go:1925 +0x636
main.main()
	_testmain.go:219 +0x19c
```

**Comment by nayihz** [comment]
We can reproduce this issue in unit testing.
```diff
diff --git a/pkg/controller/statefulset/stateful_set_control_test.go b/pkg/controller/statefulset/stateful_set_control_test.go
index f018646d3af..2592a93e2cb 100644
--- a/pkg/controller/statefulset/stateful_set_control_test.go
+++ b/pkg/controller/statefulset/stateful_set_control_test.go
@@ -2989,6 +2989,16 @@ func parallelScaleUpStatefulSetControl(set *apps.StatefulSet,
        if err != nil {
                return err
        }
+       rand.Seed(time.Now().UnixNano())
+
+       for i := 0; i < 100; i++ {
+               go func() {
+                       time.Sleep(time.Duration(rand.Int63n(201)) * time.Millisecond)
+                       if len(set.Spec.VolumeClaimTemplates) != 0 {
+                               set.Spec.VolumeClaimTemplates[0].Labels["aaaaa"] = "bbbbb"
+                       }
+               }()
+       }

        // Give up after 2 loops.
        // 2 * 500 pods per loop = 1000 max pods <- this should be enough for all test cases.
diff --git a/pkg/controller/statefulset/stateful_set_utils_test.go b/pkg/controller/statefulset/stateful_set_utils_test.go
index d5d87724dad..c99eaccdd65 100644
--- a/pkg/controller/statefulset/stateful_set_utils_test.go
+++ b/pkg/controller/statefulset/stateful_set_utils_test.go
@@ -786,6 +786,7 @@ func newPVC(name string) v1.PersistentVolumeClaim {
                ObjectMeta: metav1.ObjectMeta{
                        Namespace: v1.NamespaceDefault,
                        Name:      name,
+                       Labels:    make(map[string]string),
                },
                Spec: v1.PersistentVolumeClaimSpec{
                        Resources: v1.VolumeResourceRequirements{
```
```bash
# go test -v -run TestParallelScale
=== RUN   TestParallelScale/scale_up_from_3_to_30
 fatal error: concurrent map writes

 goroutine 381 [running]:
 k8s.io/kubernetes/pkg/controller/statefulset.getPersistentVolumeClaims(0xc0007fc500, 0x0?)
     /root/work/kubernetes/pkg/controller/statefulset/stateful_set_utils.go:347 +0x2e9
 k8s.io/kubernetes/pkg/controller/statefulset.(*StatefulPodControl).createPersistentVolumeClaims(0xc0005a5cc0, 0x6b9d92?, 0xc0004c2000?)
     /root/work/kubernetes/pkg/controller/statefulset/stateful_pod_control.go:342 +0x65
 k8s.io/kubernetes/pkg/controller/statefulset.(*StatefulPodControl).createMissingPersistentVolumeClaims(0x2442790?, {0x243e1a8, 0x365cb80}, 0x2161e51?, 0x36?)
     /root/work/kubernetes/pkg/controller/statefulset/stateful_pod_control.go:322 +0x39
 k8s.io/kubernetes/pkg/controller/statefulset.(*defaultStatefulSetControl).processReplica(0xc00064a180, {0x243e1a8, 0x365cb80}, 0xc0007fc500, 0xc0002fba40, 0xc0002fba40,     0x0?, 0xc000500d00?, 0x0, {0xc0007a22d0, ...}, ...)
     /root/work/kubernetes/pkg/controller/statefulset/stateful_set_control.go:420 +0x698
 k8s.io/kubernetes/pkg/controller/statefulset.(*defaultStatefulSetControl).updateStatefulSet.func1(0x0?)
     /root/work/kubernetes/pkg/controller/statefulset/stateful_set_control.go:626 +0x5a
 k8s.io/kubernetes/pkg/controller/statefulset.slowStartBatch.func1(0xc000788810?)
     /root/work/kubernetes/pkg/controller/statefulset/stateful_set_control.go:292 +0x5d
 created by k8s.io/kubernetes/pkg/controller/statefulset.slowStartBatch in goroutine 155
     /root/work/kubernetes/pkg/controller/statefulset/stateful_set_control.go:289 +0x15e
```

**Comment by aojea** [comment]
> Maybe it was imported by this pr https://github.com/kubernetes/kubernetes/pull/117865

The reported versions is `"1", Minor:"27", GitVersion:"v1.27.5` and it seems that PR got backported further 👀 

/priority important-soon

**Comment by aleksandra-malinowska** [comment]
The original report looks like #117865 could have introduced it, yes. I wonder if we intentionally modify template without creating a copy (and as it seems, assign the exact same labels every time we use this template for any pod): https://github.com/kubernetes/kubernetes/blob/v1.27.5/pkg/controller/statefulset/stateful_set_utils.go#L343

Regardless, that was probably safe before the pod creations were parallelized. I don't think the proposed unit test catches the bug (more on that below), but we can probably reproduce it by timing pod state transitions so that many pods are simultaneously pending and missing a PVC.

> We can reproduce this issue in unit testing.
> 
> ```diff
> diff --git a/pkg/controller/statefulset/stateful_set_control_test.go b/pkg/controller/statefulset/stateful_set_control_test.go
> index f018646d3af..2592a93e2cb 100644
> --- a/pkg/controller/statefulset/stateful_set_control_test.go
> +++ b/pkg/controller/statefulset/stateful_set_control_test.go
> @@ -2989,6 +2989,16 @@ func parallelScaleUpStatefulSetControl(set *apps.StatefulSet,
>         if err != nil {
>                 return err
>         }
> +       rand.Seed(time.Now().UnixNano())
> +
> +       for i := 0; i < 100; i++ {
> +               go func() {
> +                       time.Sleep(time.Duration(rand.Int63n(201)) * time.Millisecond)
> +                       if len(set.Spec.VolumeClaimTemplates) != 0 {
> +                               set.Spec.VolumeClaimTemplates[0].Labels["aaaaa"] = "bbbbb"
> +                       }
> +               }()
> +       }
> 
>         // Give up after 2 loops.
>         // 2 * 500 pods per loop = 1000 max pods <- this should be enough for all test cases.
> diff --git a/pkg/controller/statefulset/stateful_set_utils_test.go b/pkg/controller/statefulset/stateful_set_utils_test.go
> index d5d87724dad..c99eaccdd65 100644
> --- a/pkg/controller/statefulset/stateful_set_utils_test.go
> +++ b/pkg/controller/statefulset/stateful_set_utils_test.go
> @@ -786,6 +786,7 @@ func newPVC(name string) v1.PersistentVolumeClaim {
>                 ObjectMeta: metav1.ObjectMeta{
>                         Namespace: v1.NamespaceDefault,
>                         Name:      name,
> +                       Labels:    make(map[string]string),
>                 },
>                 Spec: v1.PersistentVolumeClaimSpec{
>                         Resources: v1.VolumeResourceRequirements{
> ```
> 
> ```shell

I ran this test on `v1.27.3` (before the cherry-pick of parallel statefulset scaling), and it also failed. This is because the test itself introduces a concurrent map write. It probably doesn't replicate the issue, and I don't think it can be merged, as it'll fail even if the original PR is reverted.

For reproduction we'd need to catch a concurrent write without adding one in test code, just as it happened based on the original report.

**Comment by aleksandra-malinowska** [comment]
Turns out we never tested this path at all, since the test PVC didn't have any labels of its own. After adding labels, existing tests are sufficient to detect race condition.

PR showing that improved test detects the issue: #121143
PR with fix and improved test: #121142

**Comment by dims** [comment]
Sweet! thanks @aleksandra-malinowska !
